### ISSUE TITLE
Question: Should an offline consumer handle an integrarion event at a later date?

### ISSUE THREAD

**Issue by w00519772** [dotnet-architecture/eShopOnContainers#822]
Thank you again for developing eShopOnContainers - it is a very helpful reference application.

I am currently reviewing the EventBusRabbitmq code: https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/BuildingBlocks/EventBus/EventBusRabbitMQ/EventBusRabbitMQ.cs

Say I have a Microservice (Microservice A) that handles an integration event (integration event A).  Microservice A is the only Microservice that handles event A.  Say microservice A is down i.e. offline (say the server is shudown) when event A is raised.  I believe this means the message is lost (even though it is supposedly durable).  Have I understood this correctly? Is there a workaround for this? i.e. to ensure that a message is received by a microservice at a later date when it is offline.

**Comment by mohammad-awad** [comment]
I have the same issue!

**Comment by w0051977** [comment]
@mohammad-awad, the issue for me was with my configuration.  The publisher and consumer were competing to handle the integration event.  

In my case; the publisher and consumer had the same 'SubscriptionClientName' in appsettings.json.  Once I ensured they have different values, then things started working as expected.

**Comment by yang-xiaodong** [comment]
@w00519772  @mohammad-awad 

The integration event in eShopOnContainers is just a reference. It doesn't take into account all the exception scenarios, but in actual business projects you need to consider the exceptions you are talking about like service shutdown or offline .

In project we need to deal with the problems, we can use the Ack mechanism provided by RabbitMQ to ensure that consumers receive messages and storaged. The example in eShop has already done this, but one scenario that is missing consideration is that if the persistence fails, the message will be lost. So we need to move the retry operations forward so they are on the publisher side or RabbitMQ.

The easiest way is to use the `BasicReject` provided by RabbitMQ to reject the message and allow them to requeue.  The consumer will receive the message again when it is online.

The above operation may still have problems, such as a server crash when `BasicReject` , But the probability is much smaller than the probability of database persistence exception.

I've provided a solution and considered most of the abnormal scenario, if you're interested, you can view it here:  https://github.com/dotnetcore/CAP

**Comment by MarcDrexler** [comment]
I'm having a similar issue. Not with RabbitMQ, but with Azure Service Bus.
My consumer is not processing any of the messages that are still waiting on the Service Bus when it is online again. As long as my consumer is online, everything is fine.
Is this a known behaviour, a bug, or am I doing something wrong?
It seems to work with RabbitMQ.

**Comment by mvelosop** [comment]
Hi @MarcDrexler,

I haven't tried with ServiceBus, but perhaps this could also be related to https://github.com/dotnet-architecture/eShopOnContainers/issues/612, in that message might get lost because the consumer is enabled before the handlers are set up, so messages might get consumed but nos handled.

I'm getting to explore this issue.

Hope this helps in the meantime.
